platform/logging/LoggerFactory.ts

```diff
@@ -0,0 +1,127 @@
+import { BufferAppender } from './appenders/buffer/BufferAppender';
+
+/**
+ * Interface that defines a way to retrieve a logger.
+ */
+export interface LoggerFactory {
+  /**
+   * Returns a `Logger` instance for the specified context.
+   * @param context Context to return logger for.
+   */
+  get(context: string): Logger;
+}
+
+/** @internal */
+export class MutableLoggerFactory implements LoggerFactory {
+  private config?: LoggingConfig;
+  private readonly appenders: Map<string, DisposableAppender> = new Map();
+  private readonly bufferAppender = new BufferAppender();
+  private readonly loggers: Map<string, LoggerAdapter> = new Map();
+
+  get(context: string): Logger {
+    if (this.loggers.has(context)) {
+      return this.loggers.get(context)!;
+    }
+
+    let loggerLevel, loggerAppenders;
+    if (this.config) {
+      const { level, appenders } = this.getLoggerConfigByContext(
+        this.config,
+        context
+      )!;
+      loggerLevel = LogLevel.fromId(level);
+      loggerAppenders = appenders.map(
+        appenderKey => this.appenders.get(appenderKey)!
+      );
+    } else {
+      // If we don't have config yet, use `buffered` appender that will store all logged messages in the memory
+      // until the config is ready.
+      loggerLevel = LogLevel.All;
+      loggerAppenders = [this.bufferAppender];
+    }
+
+    this.loggers.set(
+      context,
+      new LoggerAdapter(new BaseLogger(context, loggerLevel, loggerAppenders))
+    );
+
+    return this.loggers.get(context)!;
+  }
+
+  /**
+   * Updates all current active loggers with the new config values.
+   * @param config New config instance.
+   * @returns Promise that is resolved once all loggers are successfully updated.
+   */
+  async updateConfig(config: LoggingConfig) {
+    // Config update is asynchronous and may require some time to complete, so we should invalidate
+    // config so that new loggers will be using BufferAppender until newly configured appenders are ready.
+    // TODO: Before disposing of appenders, we should switch all existing loggers to BufferAppender first.
+    this.config = undefined;
+
+    for (const appender of this.appenders.values()) {
+      await appender.dispose();
+    }
+
+    this.appenders.clear();
+    for (const [appenderKey, appenderConfig] of config.appenders.entries()) {
+      this.appenders.set(appenderKey, Appenders.create(appenderConfig));
+    }
+
+    for (const [loggerKey, loggerAdapter] of this.loggers.entries()) {
+      const loggerConfig = this.getLoggerConfigByContext(config, loggerKey);
+      loggerAdapter.logger = new BaseLogger(
+        loggerKey,
+        LogLevel.fromId(loggerConfig.level),
+        loggerConfig.appenders.map(
+          appenderKey => this.appenders.get(appenderKey)!
+        )
+      );
+    }
+
+    this.config = config;
+
+    // Re-log all buffered log records with newly configured appenders.
+    for (const logRecord of this.bufferAppender.flush()) {
```

Comment:
I've moved `bufferAppender` to a separate property since it should be used during all config updates in the future to not miss anything (eg. file appender is being asynchronously disposed, but module would like to log something at the same time, so before closing file we replace file appender with buffer one and we should not miss anything).